shelley/chain-and-ledger/executable-spec/test/Generator/Delegation.hs

```diff
@@ -109,3 +124,47 @@ genDeRegKeyCert keys delegSt =
   where
     registered k = k ∈ dom (_stkCreds delegSt)
     availableKeys = filter (registered . toCred . snd) keys
+
+-- | Generate and return a RegPool certificate along with its witnessing key.
+genRegPool
+  :: KeyPairs
+  -> VrfKeyPairs
+  -> DPState
+  -> Gen (Maybe (DCert, KeyPair))
+genRegPool keys vrfKeys dpState =
+  if null availableKeys || null availableVrfKeys
+     then pure Nothing
+     else do
+       Just <$> genDCertRegPool availableKeys availableVrfKeys
+ where
+  notRegistered k = k `notElem` (dpState ^. pstate . pParams . to Map.elems . to (_poolPubKey <$>))
+  availableKeys = filter (notRegistered . hashKey . vKey . snd) keys
+
+  notRegisteredVrf k = k `notElem` (dpState ^. pstate . pParams . to Map.elems . to (_poolVrf <$>))
+  availableVrfKeys = filter (notRegisteredVrf . hashKeyVRF . snd) vrfKeys
+
+-- | Generate PoolParams and the key witness.
+genStakePool :: KeyPairs -> VrfKeyPairs -> Gen (PoolParams, KeyPair)
+genStakePool skeys vrfKeys = do
+  poolKeyPair   <- Gen.element skeys
+  vrfKey        <- snd <$> Gen.element vrfKeys
+  cost          <- Coin <$> genInteger 1 100
+  pledge        <- Coin <$> genInteger 1 100
+  marginPercent <- genNatural 0 100
+  acntKey       <- getAnyStakeKey skeys
+  let interval = unsafeMkUnitInterval $ fromIntegral marginPercent % 100
+  let pps = PoolParams
+              (hashKey . vKey . snd $ poolKeyPair)
+              (hashKeyVRF vrfKey)
+              pledge
+              cost
+              interval
+              (RewardAcnt $ KeyHashObj $ hashKey acntKey)
+              Set.empty
+  pure (pps, snd poolKeyPair)
+
+-- | Generate `RegPool` and the key witness.
+genDCertRegPool :: KeyPairs -> VrfKeyPairs -> Gen (DCert, KeyPair)
+genDCertRegPool skeys vrfKeys = do
+  (pps, poolKey) <- genStakePool skeys vrfKeys
+  pure (RegPool pps, poolKey)
```

Comment:
These were taken from the `Generator` and were written by @mgudemann. Unfortunately git wasn't able to keep track of that and shows that I added them...